### PR TITLE
add windows ci setting

### PR DESCRIPTION
**Pull request by yhmtsai**
~~This PR add the pipeline for windows gitlab shared runner.~~
This PR add the github action for windows
For cuda, it only test compilation. For nocuda, it test compilation and google test.

I only keep gitlab ci in commit.
gitlab ci does not give separate badge to show windows compiliation and gives 2000 min/month in our current gitlab plan.
Github action gives badge based on workflow and give unlimited min for open source.
However, it only gives 14GB for storage, so compiling windows with static and debug is failed.

github action strategy allows to evaluate all combination of matrix.
For example,
```
build_type: ["Debug", "Release"]
shared: ["ON", "OFF"]
```
gives four jobs.

I only specify only two config to reduce the list of pull request check.

~~Gitlab seems to have whole pipeline badge.
In #470 , Github action can give badges on different workflow.
Thus, I will try to see whether it also works on github action. 
By doing so, we can separate the status from different platform~~

Summary:
- MSVC: cuda (only compile), ref (build/test/install)
- MinGW: omp (build/test/install)
- Cygwin: omp (build/test/install)
- split install and compile in OSX-build

Note:
1. I do not find a way to use cygwin as a shell in github action, so I need to set cygwin path only to make everything work.
2. MSVC/mingw install Ginkgo into `C:\Program files(x86)\Ginkgo` not `C:\Program files\Ginkgo`.
I can not use `CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT` to set prefix force, which leads that user can only set `CMAKE_INSTALL_PREFIX` in the beginning. reference: https://gitlab.kitware.com/cmake/cmake/issues/18312
3.  MSVC nvcc has flags like `-Xcompiler="-O2 ..."`, so I add a macro to modify cuda_flag for GinkgoConfig.pc by `"` -> `\"`
4. Github action create a script containing the commands in the `run:` and use shell to run the script, so github action only catch the last error code. For example, in [workflow](https://github.com/ginkgo-project/ginkgo/runs/493420377?check_suite_focus=true), the parilu test is failed in configure step, but the last command `make install` is successful. Thus, the configure is successful.

TODO:
- [x] Merge #473 
- [x] MSVC env (for CUDA)
- [x] MinGW/CYGWIN env (for OpenMP)
- [x] MinGW/CYGWIN test_install
- [x] MSVC test_install
- [x] Ginkgo.pc warning

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -799,3 +799,93 @@ new-issue-on-failure:
...
+# Only compile cuda and reference.
+compile/cuda102/msvc2019/cuda/release/shared:
+  <<: *windows_default_ci
+  variables:
+    <<: *windows_default_variables
+    BUILD_CUDA: "ON"
+    BUILD_TYPE: Release
+  before_script:
+    - .windows_default_before_script
+    - curl.exe -L http://developer.download.nvidia.com/compute/cuda/10.2/Prod/local_installers/cuda_10.2.89_441.22_win10.exe --output cuda.exe
```
Comment (upsj) [inline .gitlab-ci.yml]:
```suggestion
    - curl.exe -L https://developer.download.nvidia.com/compute/cuda/10.2/Prod/local_installers/cuda_10.2.89_441.22_win10.exe --output cuda.exe
```
for security reasons :)

**Comment by pratikvn** [comment]
It seems that OpenMP [is being detected](https://github.com/ginkgo-project/ginkgo/pull/471/checks?check_run_id=484435904#step:5:30), but you do not compile for it. Is that for any specific reason  ?

**Comment by yhmtsai** [comment]
MSVC only supports OpenMP 2, but we use something unavailble in OpenMP 2

File: .github/workflows/windows-build.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        echo "::add-path::$pwd\build\windows_shared_library"
+    - name: debug_static_env
+      if: matrix.config.shared == 'OFF' && matrix.config.build_type == 'Debug'
+      run: |
+        echo "::set-env name=CXXFLAGS::/bigobj"
+    - name: configure
+      run: |
+        mkdir build
+        cd build
+        cmake .. -DBUILD_SHARED_LIBS=${{ matrix.config.shared }} -DGINKGO_DEVEL_TOOLS=OFF -DGINKGO_BUILD_CUDA=OFF -DGINKGO_BUILD_OMP=OFF
```
Comment (upsj) [inline .github/workflows/windows-build.yml]:
```suggestion
        cmake .. -DBUILD_SHARED_LIBS=${{ matrix.config.shared }} -DGINKGO_BUILD_CUDA=OFF -DGINKGO_BUILD_OMP=OFF
```

File: .github/workflows/windows-build.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        choco install cuda -y
+    - name: configure
+      run: |
+        $env:ChocolateyInstall = Convert-Path "$((Get-Command choco).Path)\..\.."   
+        Import-Module "$env:ChocolateyInstall\helpers\chocolateyProfile.psm1"
+        refreshenv
+        mkdir build
+        cd build
+        $env:PATH="$pwd\windows_shared_library;$env:PATH"
+        cmake -DGINKGO_DEVEL_TOOLS=OFF -DGINKGO_BUILD_CUDA=ON -DGINKGO_BUILD_OMP=OFF ..
```
Comment (upsj) [inline .github/workflows/windows-build.yml]:
```suggestion
        cmake -DGINKGO_BUILD_CUDA=ON -DGINKGO_BUILD_OMP=OFF ..
```

File: .github/workflows/windows-build.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      if: matrix.config.shared == 'OFF' && matrix.config.build_type == 'Debug'
+      run: |
+        echo "::set-env name=CXXFLAGS::/bigobj"
+    - name: configure
+      run: |
+        mkdir build
+        cd build
+        cmake .. -DBUILD_SHARED_LIBS=${{ matrix.config.shared }} -DGINKGO_BUILD_CUDA=OFF -DGINKGO_BUILD_OMP=OFF
+        cmake --build . -j8 --config ${{ matrix.config.build_type }}
+        ctest . -C ${{ matrix.config.build_type }}
```
Comment (pratikvn) [inline .github/workflows/windows-build.yml]: Can you also add a `make install && make test_install` here ?

File: .github/workflows/windows-build.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        choco install cuda -y
+    - name: configure
+      run: |
+        $env:ChocolateyInstall = Convert-Path "$((Get-Command choco).Path)\..\.."   
+        Import-Module "$env:ChocolateyInstall\helpers\chocolateyProfile.psm1"
+        refreshenv
+        mkdir build
+        cd build
+        $env:PATH="$pwd\windows_shared_library;$env:PATH"
+        cmake -DGINKGO_BUILD_CUDA=ON -DGINKGO_BUILD_OMP=OFF ..
```
Comment (tcojean) [inline .github/workflows/windows-build.yml]: There is no way to test OpenMP on Windows?

Comment (yhmtsai) [reply .github/workflows/windows-build.yml]: I add mingw/cygwin to test OpenMP

File: .github/workflows/windows-build.yml (hunk 4 above)
Comment (yhmtsai) [reply .github/workflows/windows-build.yml]: Done

File: CMakeLists.txt [hunk 6]
```diff
@@ -229,18 +230,32 @@ endif()
...
-if (GINKGO_BUILD_CUDA)
-    set(GINKGO_TEST_INSTALL_COMMAND "${GINKGO_TEST_INSTALL_COMMAND}" && "${Ginkgo_BINARY_DIR}/test_install/test_install_cuda")
-endif()
+if (MSVC)
+    # Set path/command with $<CONFIG>
+    set(GINKGO_TEST_INSTALL_COMMAND "${Ginkgo_BINARY_DIR}/test_install/$<CONFIG>/test_install")
+    if (GINKGO_BUILD_CUDA)
+        set(GINKGO_TEST_INSTALL_COMMAND "${GINKGO_TEST_INSTALL_COMMAND}" "${Ginkgo_BINARY_DIR}/test_install/$<CONFIG>/test_install_cuda")
+    endif ()
+    set(GINKGO_BUILD_TEST_INSTALL_COMMAND COMMAND ${CMAKE_COMMAND} --build ${Ginkgo_BINARY_DIR}/test_install --config $<CONFIG>)
```
Comment (upsj) [inline CMakeLists.txt]: unquoted strings
```suggestion
    set(GINKGO_BUILD_TEST_INSTALL_COMMAND COMMAND "${CMAKE_COMMAND}" --build "${Ginkgo_BINARY_DIR}/test_install" --config $<CONFIG>)
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -229,18 +230,32 @@ endif()
...
+else ()
+    set(GINKGO_TEST_INSTALL_COMMAND "${Ginkgo_BINARY_DIR}/test_install/test_install")
+    if (GINKGO_BUILD_CUDA)
+        set(GINKGO_TEST_INSTALL_COMMAND "${GINKGO_TEST_INSTALL_COMMAND}" "${Ginkgo_BINARY_DIR}/test_install/test_install_cuda")
+    endif ()
+    set(GINKGO_BUILD_TEST_INSTALL_COMMAND COMMAND ${CMAKE_COMMAND} --build ${Ginkgo_BINARY_DIR}/test_install)
```
Comment (upsj) [inline CMakeLists.txt]:
```suggestion
    set(GINKGO_BUILD_TEST_INSTALL_COMMAND COMMAND "${CMAKE_COMMAND}" --build "${Ginkgo_BINARY_DIR}/test_install")
```

**Comment by yhmtsai** [comment]
Adding `""` is failed in cygwin. in https://github.com/ginkgo-project/ginkgo/actions/runs/52976212
cmake `command` can handle the variable containing space.
(related ref: https://stackoverflow.com/questions/8925396/why-does-cmake-prefixes-spaces-with-backslashes-when-executing-a-command/8933997#8933997)
cygwin considers "b\ uild" as "b/ uild", so it is failed

The current code has no problem with space ("b uild") on https://github.com/ginkgo-project/ginkgo/actions/runs/53528009
I keep testing "build" not "b uild" in the PR.

File: .github/workflows/osx.yml [hunk 8]
```diff
@@ -28,5 +28,8 @@ jobs:
         cmake .. -DBUILD_SHARED_LIBS=${{ matrix.config.shared }} -DCMAKE_BUILD_TYPE=${{ matrix.config.build_type }}
         make -j8 
         ctest -j10
+    - name: test_install
```
Comment (tcojean) [inline .github/workflows/osx.yml]: You name this step `install` in the Windows step. Either are good to me, but maybe keep only one for uniformity.

File: CMakeLists.txt [hunk 9]
```diff
@@ -229,18 +230,32 @@ endif()
...
+    if(GINKGO_BUILD_CUDA)
+        set(GINKGO_TEST_INSTALL_COMMAND "${GINKGO_TEST_INSTALL_COMMAND}" "${Ginkgo_BINARY_DIR}/test_install/$<CONFIG>/test_install_cuda")
+    endif()
+    set(GINKGO_BUILD_TEST_INSTALL_COMMAND COMMAND ${CMAKE_COMMAND} --build ${Ginkgo_BINARY_DIR}/test_install --config $<CONFIG>)
+else()
+    set(GINKGO_TEST_INSTALL_COMMAND "${Ginkgo_BINARY_DIR}/test_install/test_install")
+    if(GINKGO_BUILD_CUDA)
+        set(GINKGO_TEST_INSTALL_COMMAND "${GINKGO_TEST_INSTALL_COMMAND}" "${Ginkgo_BINARY_DIR}/test_install/test_install_cuda")
+    endif()
+    set(GINKGO_BUILD_TEST_INSTALL_COMMAND COMMAND ${CMAKE_COMMAND} --build ${Ginkgo_BINARY_DIR}/test_install)
```
Comment (tcojean) [inline CMakeLists.txt]: Do we have to keep this separate? We could maybe also use the `$<CONFIG>` version for Linux, or is there an issue I'm not aware of?

Comment (yhmtsai) [reply CMakeLists.txt]: using `--config` is fine in Linux. cmake ignore the `--config` option
`$<CONFIG>` is the same as `${CMAKE_BUILD_TYPE}` even if using `--config`

I am also fine with adding  `--config` and some comment like `--config is ignored in single-configuration generator` if it is clear enough.

Comment (tcojean) [reply CMakeLists.txt]: Great thanks!

**Comment by yhmtsai** [comment]
We have many badges now and they can not fit in one line.
I split them into two parts.

image

**Comment by tcojean** [comment]
@yhmtsai Thanks. I was thinking something similar, but maybe the other way around: split after reliability badge, this way first line is all code quality things, second line is information on the software (CDash dashboard, etc).

**Comment by yhmtsai** [comment]
image
This is the new layout

**Pull request merged**